// gitpad/gitpad.js
(function() {
  "use strict";

  var _git = require("gift");
  var _repo = null;
  var _valid_repo = false;

  /*
    List of functionalities to support
    1. Commit changes to local repo
    2. Push changes to remote repo
    3. Browse local repo for commits
    4. Roll back to a certain commit

  */

  //
  //===== Initialize git repository =====
  //
  exports.init = function(path) {
    if (path) {
      _repo = _git(path);

    } else {
      _repo = _git(".");

    }

    _repo.status(function(err, status) {
      if (err) {
        // Init failed, indicate error
        console.log("Git repo initialization failed: " + err);
        _valid_repo = false;
        return false;

      } else {
        // Init successful, sets flag
        console.log("Git repo initialization successful!");
        _valid_repo = true;
        return true;
      }
    })
  };


  //
  //===== Show repo status (clean? tracked/untracked files?)
  //
  exports.showStatus = function() {
    if (!_valid_repo) {
      // Did not properly init yet, return error
      return false;
    }

    _repo.status(function(err, status) {
      if (err) 
      console.log(status);
      console.log(err);
    })
  }


  //
  //===== Document version control
  //

  /*
    Lists the history of a certain file
    - filename: name of the file being queried
    - limit: an integer indicating how many commits to be shown, put "all" if want to see all commits
    TODO: add callback
  */
  exports.showFileHistory = function(filename, limit) {
    _repo.file_history(filename, limit, function(err, commits) {
      console.log(commits);
    });
  }

  /*
    Lists most recent commits applied to the whole repo (branch is default to "master")
    - limit: how many commits to return
    - skip(optional): how many commits to skip (for pagination)
  */
  exports.showHistory = function(limit, skip) {
    _repo.commits("master", limit, skip, function(err, commits) {
      if (err) {
        console.log(err);

      } else {
        console.log(commits);
      }
    });
  }

  /*
    Add the file to be saved into current commit and commit the file
    - filename: name of the file to be saved
    - msg: a message associated to the save action
  */
  exports.saveFile = function(filename, msg) {
    _repo.add(filename, function(err) {
      if (err) {
        console.log(err);
        return;
      }

      _repo.commit(msg, function(err) {
        if (err) {
          console.log(err);
          return;
        }
      });
    });
  }

  /*
    Publish selected file(s) (pushes selective file(s) from local repo to remote repo)
    - files: name of the list of files to be published
  */
  exports.publishFiles = function(files) {
    var fileCommits = [];
    // Get all the commit ids needing to be cherry-picked
    for (var i=0; i<files.length; i++) {
      _repo.file_history(files[i], 1, function(err, commits) {
        if (err) {
          console.log(err);
          return;
        }

        fileCommits.push(commits[0].id);
      });
    }

    // Switch to staging branch -> cherry pick all the commits -> switch back to dev branch
    _repo.checkout("newbranch", function(err) {
      // if (err) {
      //   console.log(err);
      //   return;
      // }
      // for (var i=0; i<fileCommits.length; i++) {
      //   _repo.cherrypick(fileCommits[i], {"strategy": "recursive", "strategy-option": "theirs"}, function(err){
      //     if (err)
      //       console.log(err);
      //   });
      // }
      _repo.checkout("master");
    });

  }

}).call(this);
